zerver/lib/user_topics.py

```diff
@@ -173,26 +176,40 @@ def bulk_set_user_topic_visibility_policy_in_database(
             )
             continue
         # The request is to just 'update' the visibility policy of a topic
-        user_profiles_seeking_visibility_policy_update.append(row.user_profile)
-
-    if user_profiles_seeking_visibility_policy_update:
-        rows.filter(user_profile__in=user_profiles_seeking_visibility_policy_update).update(
-            visibility_policy=visibility_policy, last_updated=last_updated
+        user_profiles_seeking_user_topic_update_or_create.append(row.user_profile)
+
+    if user_profiles_seeking_user_topic_update_or_create:
+        # TODO: Replace this raw SQL query with bulk_create() queryset method.
+        # Currently, django doesn't support using function call in 'unique_fields'.
+        # See #26461 and https://code.djangoproject.com/ticket/34943 for details.
+        values = SQL(", ").join(
+            [
+                SQL(
+                    "({user_profile_id}, {stream_id}, {recipient_id}, {topic_name}, {last_updated}, {visibility_policy})"
+                ).format(
+                    user_profile_id=Literal(user_profile.id),
+                    stream_id=Literal(stream_id),
+                    recipient_id=Literal(recipient_id),
+                    topic_name=Literal(topic_name),
+                    last_updated=Literal(last_updated),
+                    visibility_policy=Literal(visibility_policy),
+                )
+                for user_profile in user_profiles_seeking_user_topic_update_or_create
+            ]
```

Comment:
It occurs to me that only one column here is actually varying.  For large user id lists, this is going to be very inefficient over the wire, since we're repeating the same values for each row.

A better construction would be:
```sql
INSERT INTO zerver_usertopic(user_profile_id, stream_id, recipient_id, topic_name, last_updated, visibility_policy)
  SELECT * FROM UNNEST(ARRAY[1,2,3,4])
    users (user_profile_id)
  CROSS JOIN (VALUES (42, 17, 'whee', '2024-06-09 12:34:56'::timestamp, 4))
    other_values(stream_id, recipient_id, topic_name, last_updated, visibility_policy)
  ON CONFLICT (user_profile_id, stream_id, lower(topic_name)) DO UPDATE SET
    last_updated = EXCLUDED.last_updated,
    visibility_policy = EXCLUDED.visibility_policy;
```

We're using `CROSS JOIN` to get the full cartesian product, `UNNEST` to turn an array into a list of rows, and a hard-coded `VALUES` to make a pretend table with one row.

This does mean that the comment about "oh, we can move this with `bulk_create` is likely not accurate anymore, since we're using a more complicated construction which Django can't express at all.  So maybe we can do away with that comment and note in the commit message.

Sorry for not considering this in the previous review!

Reply:
Updated the code with this approach.

Good to know this efficient approach.